### ISSUE TITLE
Average momentum spread per particle from ParticleBinning

### ISSUE THREAD

**Issue by Nikolai-mephi** [SmileiPIC/Smilei#536]
Dear Smilei community,

I have a question on calculation of the momentum spread, by which I mean `<(p - <p>)^2>` averaged over all particles in a particular spatial bin, with the ParticleBinning diagnostic.
Initially, in my 2D simulation I tried to save three separate ParticleBinning diagnostics with three different deposited quantities:
```
DiagParticleBinning(
    deposited_quantity = 'weight',
    every = 500,
    time_average = 1,
    species = ['eon'],
    axes = [['x', 0., Lx, 236],
            ['y', 0., Ly, 268]]
)

DiagParticleBinning(
    deposited_quantity = mean_p,
    every = 500,
    time_average = 1,
    species = ['eon'],
    axes = [['x', 0., Lx, 236],
            ['y', 0., Ly, 268]]
)

DiagParticleBinning(
    deposited_quantity = mean_sqr_p,
    every = 500,
    time_average = 1,
    species = ['eon'],
    axes = [['x', 0., Lx, 236],
            ['y', 0., Ly, 268]]
)
```

where the 'deposited_quantity' functions in the second and third one were defined as follows:

```
def mean_p(particles):
    return particles.weight * np.sqrt(particles.px * particles.px + particles.py * particles.py + particles.pz * particl
es.pz)

def mean_sqr_p(particles):
    return particles.weight * (particles.px * particles.px + particles.py * particles.py + particles.pz * particles.pz)
```

I also defined the same three diagnostics for ions in my simulation. Afterwards, during post-processing, I divided the results of the third and second diagnostic by weight and subtracted the square of the second one from the third one. 

Judging from the results I obtained I can conclude that I did something wrong, as they somewhat contradict my expectations. I irradiated a plasma object with `I>10^22 W/cm2` laser and expected `~ 10-100 MeV` electrons and `~10 keV` ions, but when I used the momentum spreads to calculate the 'temperatures' as `T=sqrt(dp**2 * c**2 + m * c**2) - m * c**2`, I got a maximum of 20 MeV for electrons and less than one eV for ions, with the latter being particularly suspicious. I also tried to use 'weight_p' in the second ParticleBinning and obtained another, yet still incorrect result - in this case most of the values of `<p2> - <p>**2` for ions were negative.

Could it be that I misunderstood the meaning of weights or how particle binning diagnostic works in general? 

In this case, what would be the correct way to implement particle binning diagnostics in order to calculate average momentum spread per particle and the temperature for this sort of particle using the definition above?

**Comment by mccoys** [comment]
You have understood weights correctly it seems.  But the particles.px quantities are misleading as they are in fact gammax × vx. The mass is missing.  It could explain the discrepancy with ions

**Comment by Nikolai-mephi** [comment]
Thank you for your response, @mccoys !

Unfortunately, I was still unable to resolve my issue. After substituting electron mass with ion mass for the ion case, I, conversely, obtained too high values for ions. Their temperature defined with the formula above exceeds electron temperature by a factor of 30-40. That is quite strange considering that the only way for a particle to get energy in my simulation is through fields, and electrons supposedly should get it more effectively due to much higher charge to mass ratio.

Do you have an idea what can cause such discrepancy?

**Comment by mccoys** [comment]
It is very hard to find a solution without more details. Are you sure your expected values are valid in average?

**Comment by Nikolai-mephi** [comment]
You are right. I am not completely sure in the expected values. Maybe I should try a much simpler example to check the validity of my implementation of the ParticleBinning diagnostic - define plasma with preset electron and ion temperatures and maxwell-juttner distributions, and try to see what I obtain with the same diagnostic but without any laser. If the approach is correct, than at the start of the simulation I will see values that will be similar to the initially set temperatures.

**Comment by Nikolai-mephi** [comment]
So I have done a simplified simulation with a uniform circle plasma object with the initial electron temperature of 5 MeV and initial ion temperature of 2.5 keV. Here's the namelist for it.
[namelist.txt](https://github.com/SmileiPIC/Smilei/files/8696102/namelist.txt)

The initial electron temperature seems to correspond quite well to the set value, while the ion temperature appears somewhat lower - below 1 keV while it should be at around 2.5 keV.
[Te_0000000000.pdf](https://github.com/SmileiPIC/Smilei/files/8696105/Te_0000000000.pdf)
[Ti_0000000000.pdf](https://github.com/SmileiPIC/Smilei/files/8696106/Ti_0000000000.pdf)

Of course, it is not such a great discrepancy as in my main simulation, but I am starting to think I should indeed reconsider the expected values. Maybe by the time I look at the temperatures in the main sim all hot electrons escape the simulation box while ions remain there due to higher mass and accumulate higher temperature.

However, I am still curious to learn why the initial ion temperature is slightly lower than the set value. Is it how it should be and I am just missing something?

Here's the script I use for post-processing and where I calculate the temperatures.
[plotter.txt](https://github.com/SmileiPIC/Smilei/files/8696111/plotter.txt)

**Comment by mccoys** [comment]
A suggestion: temperature is not defined as `<v^2> - <v>^2`,  but with vectors like ![eq](http://www.sciweavers.org/tex2img.php?eq=%5Cleft%3C%28%5Cvec%20v%20-%20%5Cleft%3C%5Cvec%20v%5Cright%3E%29%5E2%20%5Cright%3E%20%3D%20%5Cleft%3Cv%5E2%5Cright%3E%20-%20%5Cleft%3C%5Cvec%20v%5Cright%3E%5E2&bc=White&fc=Black&im=jpg&fs=12&ff=arev&edit=0). Your last term is wrong as you use <v> instead of ![eq](http://www.sciweavers.org/tex2img.php?eq=%5Cleft%3C%5Cvec%20v%5Cright%3E&bc=White&fc=Black&im=jpg&fs=12&ff=arev&edit=0)

**Comment by Nikolai-mephi** [comment]
Thank you! I think I understood your point, but may I ask you to send the equation againg in some other format? I cannot open the current one properly due to 'Non image content returned' error.

**Comment by Nikolai-mephi** [comment]
Greetings!

I am still trying to figure the reason for the consistency in my original simulation. I have modified my simulation and now average px, py and pz are saved separately and I subtract the squared momentum averaged as a vector rather than an absolute value. I have done some more tests with the same simplified simulation I have referenced previously and for some reason, I cannot obtain the same results if I add drift velocity to all particles. I expected that if calculate squared momentum in the frame of reference with zero average momentum, the results should be the same no mater what drift velocity I add, however they are not and in the case of non-zero drift velocity they are higher.
[Te_without_drift.pdf](https://github.com/SmileiPIC/Smilei/files/8750305/Te_without_drift.pdf)
[Te_with_drift.pdf](https://github.com/SmileiPIC/Smilei/files/8750306/Te_with_drift.pdf)
For ions there also seems to be a difference by a factor of ~3.

I suspect that in my main simulation it might be something similar - drift velocity is not subtracted properly and is added to the results, leading to higher values.

Here is the modified files for the cases of zero and non-zero drift velocities and the script I use for post-processing. 
[namelist1.txt](https://github.com/SmileiPIC/Smilei/files/8750307/namelist1.txt)
[namelist2.txt](https://github.com/SmileiPIC/Smilei/files/8750308/namelist2.txt)
[plotter.txt](https://github.com/SmileiPIC/Smilei/files/8750309/plotter.txt)

May I ask you to have a look and tell me what I am still doing wrong?

**Comment by mccoys** [comment]
Did you correct velocities with a lorentz boost?

**Comment by Nikolai-mephi** [comment]
No, I did not multiply by the Lorenz boost. It was not my goal to find the velocity spread (which I assume the Lorentz is needed for) but to find the momentum spread by computing the average square momentum in the frame of reference where the average momentum is zero. If I move into this frame of reference by subtracting `<px>`, `<py>`, `<pz>` from the relevant `px`, `py`, `pz` components, this results in the subtraction of `<px>^2`, `<py>^2` and `<pz>^2` from the `<p^2>` value in the equation for the squared momentum spread. The momentum spread computed in this way should be independent on the average particle motion with some drift velocity or momentum - this is what I initially thought. Have I made a mistake at some point in my assumptions?

**Comment by mccoys** [comment]
If your particles are relativistic, you cannot simply use the usual Maxwellian distribution.

**Comment by Nikolai-mephi** [comment]
Yes,  understand. In this case I will reformulate my question - if i define species with maxwell-juttner distributions and some temperature, what is the correct way, if any, to obtain back this temperature from particle binning in the case, when there is also some relativistic drift velocity?

**Comment by mccoys** [comment]
There is no Lorentz boost directly included in the particleBinning diagnostic. Instead, you could make a new custom axis `boosted_px` that transforms all px to their value in the frame where the distribution is at rest.

**Comment by Nikolai-mephi** [comment]
Thank you for your suggestions. I was finally able to obtain the same results for the 'zero drift velocity' and 'non-zero drift velocity' cases. Eventually, I ended up with 28 ParticleBinning diagnostics (different combinations of E, px, py, pz multiplied by one another and by weights, as well as weights themselves for electrons and ions) which I had to write to use the Lorenz boost matrix and calculate `p^2` in the frame which may move along all 3 axes. The resulting `p^2` values in this frame seem to be more or less independent on the drift velocity.
[Ti_no_drift.pdf](https://github.com/SmileiPIC/Smilei/files/8825065/Ti_no_drift.pdf)
[Ti_drift.pdf](https://github.com/SmileiPIC/Smilei/files/8825066/Ti_drift.pdf)

Now I need to check what I will obtain in my main simulation.
